notebooks/community/aisamples/AIsample - Uplift Modelling.ipynb

```diff
@@ -0,0 +1,882 @@
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "treatment_train_df = train_df.where(f\"{TREATMENT_COLUMN} > 0\")\n",
+    "control_train_df = train_df.where(f\"{TREATMENT_COLUMN} = 0\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "## Step 3: Model Training and Evaluation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "### Uplift Modelling: T-Learner with LightGBM"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "def train(train_df):\n",
+    "    classifier = (\n",
+    "        LightGBMClassifier()\n",
+    "        .setFeaturesCol(\"features\")\n",
+    "        .setNumLeaves(10)\n",
+    "        .setNumIterations(100)\n",
+    "        .setObjective(\"binary\")\n",
+    "        .setLabelCol(LABEL_COLUMN)\n",
+    "    )\n",
+    "\n",
+    "    model = classifier.fit(train_df)\n",
+    "\n",
+    "    return model\n",
+    "\n",
+    "\n",
+    "treatment_model = train(treatment_train_df)\n",
+    "control_model = train(control_train_df)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "### Predict on test dataset"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "getPred = F.udf(lambda v: float(v[1]), FloatType())\n",
+    "\n",
+    "test_pred_df = (\n",
+    "    treatment_model.transform(test_df)\n",
+    "    .withColumn(\"treatment_pred\", getPred(\"probability\"))\n",
+    "    .drop(\"rawPrediction\", \"probability\", \"prediction\")\n",
+    ")\n",
+    "\n",
+    "test_pred_df = (\n",
+    "    control_model.transform(test_pred_df)\n",
+    "    .withColumn(\"control_pred\", getPred(\"probability\"))\n",
+    "    .drop(\"rawPrediction\", \"probability\", \"prediction\")\n",
+    ")\n",
+    "\n",
+    "test_pred_df = test_pred_df.withColumn(\n",
+    "    \"lift_pred\", F.col(\"treatment_pred\") - F.col(\"control_pred\")\n",
+    ").select(TREATMENT_COLUMN, LABEL_COLUMN, \"treatment_pred\", \"control_pred\", \"lift_pred\")\n",
+    "\n",
+    "display(test_pred_df.limit(20))"
```

Comment:
nit: this can be simplified alot (one single dot chain) with out "FluentAPI" See usage here:
https://github.com/microsoft/SynapseML/blob/3bd5bef91db9339f4b7db6b6a800b08f68ed25e8/website/versioned_docs/version-0.9.5/features/responsible_ai/Interpretability%20-%20Snow%20Leopard%20Detection.md